### ISSUE TITLE
Training DirectedSAGE node classification with multiple graphs and partly labeled nodes (such as PPI Dataset)

### ISSUE THREAD

**Issue by imayachita** [stellargraph/stellargraph#1881]
Hi,
I would like to ask how can I train DirectedSAGE node classification with multiple graphs? The example of a similar setting is Protein-Protein Interaction Dataset. 
![image](https://user-images.githubusercontent.com/3615586/109399246-4c684080-7974-11eb-824f-de0125fd95e6.png)
(reference: [GAT](https://arxiv.org/pdf/1710.10903.pdf) paper)

For example, I have 15 graphs, where I want to use the first 10 graphs as the training data and the other 5 graphs as the val data. Do I need to use [union](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.operators.binary.union.html#networkx.algorithms.operators.binary.union) to combine all 15 graphs in a giant graph or is there any other workaround? Is it possible to have different node generators for train and val? The nodes are partly labeled (there are nodes whose labels are None). How can I set up my generator and training? 

And also, may I get more explanation about `in_samples` and `out_samples`? I don't really get what they represent.

Thanks

**Comment by huonw** [comment]
> Do I need to use union to combine all 15 graphs in a giant graph or is there any other workaround?

Yep, I think something along those lines will be the best way. I think you may also be able to just union the 10 training graphs together, and the 5 val graphs together, resulting in two large graphs.

> Is it possible to have different node generators for train and val?

I think so. Have you tried it and it failed?

> How can I set up my generator and training?

I may not understand the question, but the best way to handle this is to only pass labelled nodes into the `.flow` call; this ensures that the model is only trained against those nodes. Does that make sense?

> And also, may I get more explanation about in_samples and out_samples? I don't really get what they represent.

Suppose there's a graph that contains edges like the following (where each row of the table represents an edge like `source -> target`):

| source |  target |
|---|---|
| N | a |
| N | b |
| N | c |
| d | N |
| e | N |

In undirected GraphSAGE, the `num_samples` parameter to the generator represents the size of the neighbour sampled at each step. That is, the number of edges selected and traversed to choose the nodes in the next layer of the model. For instance, if the current head node is N, the first layer's nodes will be selected from a, b, c, d, or e (with replacement).

In directed GraphSAGE, the same process happens, except edges independently followed both forward and backwards:

- `in_samples`: the number of edges traversed backwards, meaning edges where the current head node is the target. For instance, if it's N, this first layer's in samples will be selected from d or e.
- `out_samples`: the number of edges traversed forwards, meaning edges where the current head node is the target. For instance, if it's N again, the out samples will be selected from a, b or c.

For instance, with `in_samples = [1, 3, 5]` and `out_samples = [2, 4, 6]`, the first layer might sample `[d]`, `[a, b]`. The next layer will then sample 3 in nodes and 4 out nodes for each of these (so 9 in and 12 out total in this layer), and then the third layer will sample 5 in nodes and 6 out nodes for each of those (so (9 + 12) &times; 5 = 105 in and (9 + 12) &times; 6 = 126 out total in this layer).

Each of these will be aggregated in the same way as conventional GraphSAGE, except it will be using separate weight matrices for the in nodes and the out nodes. This allows the model to learn different information for the head node (the focus of the aggregation) based on the direction of the connection to the sampled node.

---

Does that help?

(Thanks for waiting!)

**Comment by imayachita** [comment]
Thanks a lot @huonw!

> Yep, I think something along those lines will be the best way. I think you may also be able to just union the 10 training graphs together, and the 5 val graphs together, resulting in two large graphs.

My concern is that I have a lot of large graphs that cause memory problems when I union all of them. Is there any more efficient way to iterate on those graphs? For example, in addition to node generator, making it graph generator instead?
